### ISSUE TITLE
Use Proxy with Spfi and Graphi, proxy url was injected in payload of batched request

### ISSUE THREAD

**Issue by MozzaikJD** [pnp/pnpjs#2893]
### What version of PnPjs library you are using

3.x

### Minor Version Number

20.1

### Target environment

Browser App (Hosted external to Microsoft 365 platform)

### Additional environment details

Use local proxy to add authentication and add some behaviors (metrics)

### Question/Request

Hello pnpJS community,

I am trying to use an reverse proxy to pass my graph and sharepoint request, for example "http://localhost:3333/_api".
I setup my proxy to redirect some root :
- _api/graph to https://graph.microsoft.com/v1.0/
- _api/sharepoint to https://mytenant.sharepoint.com/_api/

I am using behavior to add my proxy url to my request :
Graph : behaviors.push(GraphBrowser({ baseUrl: PROXY_URL_GRAPH }));
Sharepoint behaviors.push(SPBrowser({ baseUrl: PROXY_URL }));

This is working fine for simple graph or sharepoint request.
But for batched request, I retrieve my proxy url in the body request who was sent to the target api.
Obviously, the target api doesn't know my proxy url, so I get an error.

For graph, I find a trick to remove the proxy url from the body request to keep only the relative url (with observer), but I think it's not the best way to do it.
For sharepoint, it use the full url, I tried a similar trick to replace the proxy url by the direct sharepoint url, but it doesn't work.

I didn't find any way to configure the proxy url for batched request (or the correct way to do it), so I am asking you if you have any idea to do it.

Thank you!

**Comment by bcameron1231** [comment]
AFAIK, this isn't going to be possible. While Graph does allow the relative resource URL in the individual request... SharePoint batching endpoints does not. SharePoint batching Request URL and the URLs in the multipart body **must contain** the root service URL and they must match the Request URL.

**Comment by patrick-rodgers** [comment]
To add to what Beau says - this isn't something we're going to invest in supporting through the library as it is outside our core use cases.

**Comment by MozzaikJD** [comment]
Thank you for your answer.
I understand that it is not a core use case, but I investigated a little bit more and I think I found a solution.
I encountered another problem with the batched request, but I think it's not related to the proxy.
I will open another issue for this problem if I confirm that it's not related to the proxy.

Below, I explain how I did it, if someone else is interested :


I keep the baseUrl of the proxy in the behaviors, but I don't use it for the batched request.
To make it work, I pass the SharepointUrl to my job service (who is responsible to manage pnplib request)
I build normally my SPFI object and I get a batched version by this call 
const [batchedSP, execute] = sp.batched();
After that I replace my browser behavior by a new one with the sharepoint url.
const batchedSP.using(SPBrowser({ baseUrl: sharepointUrl }));

With this trick :
 - My global batched request is sent to the proxy, 
 - but the body request refer to the sharepoint url.
I don't know if it's the best way to do it, but it's working for me.

I hope it will help someone else in the future.

Quickly, I explain my other problem : I use the batchedSP to get a details of set in term store and the terms of the term set.
In my first try, the batched request was sent to the proxy, but the api version was not present in the url, but in the body request, termstore introspection use the 2.1 version of the api, so I got an error.
"The expression "v2.1/termstore/sets/8ed8c9ea-7052-4c1d-a4d7-b9c10bffea6f" is not valid."
To fix it, I add the api version directly on the proxy side.
After that, I got another error, the parsing of the body request failed.
I identified the problem, the body request was sent in json format. But for SP, the body request must be in application/http format.
To fix it, I add a new header to the request "Content-Type: application/http" with the following code :
sp.using(InjectHeaders({ "Content-Type": "application/http" }));
After that, it's working fine ... but with a lot of tricks...

I will try to check if I reproduce these problems without the proxy, to be sure that it's not related to it.

Thank again for your answer and your work on this library.
